### PR TITLE
Auto-fetch updates

### PR DESCRIPTION
**Pull request by Focshole**
I have finished writing the code and it seems to work, now tests are fine, test coverage is around 81%. There are no breaking changes with the original version, with a original version's config file, it would run exactly as before and viceversa. No breaking changes with configs.
The only change that may confuse you is that the function "downloadFile" got renamed to "downloadFileAndSend", cause it does both. Now downloadFile simply downloads a file (I use it for prefetching).

I have added a sqlite3 database to keep track of the downloaded databases links, integrated with GORM. Can be hard to debug issues with GORM, especially because it is not strongly typed and because queries use strings (which provide no hints if some parameters are not being used) and can silently fail (with requests not found error message, which happens also at every insert).
I'll try to explain by examples GORM's (default) naming conversion which is like:
-  `TableName`->`table_name`
- `ATTRIBUTE`->`attribute`
- `AttRiBuTe`->`att_ri_bu_te`
It should be intuitive, but I wasted about half an hour on that, I don't want to waste your time too.

The database could be useful for deploying the "stats" thing written as a todo.

I'll try to describe the prefetching mechanism as how I have implemented it right now:

-  When a .db file is requested and downloaded, I save the link and the last time it got requested in the db (MirrorDB table),  db links are hard to generate otherwise.
- When a package file whose name is structured as name-version-subversion-arch.pkg.tar.zst is requested, I save name, version (merging version-subversion into version), arch and the local repo name in the db (Package table). I also keep track of when it got prefetched and when got downloaded.
- When the prefetching routine is being called, as first I delete "old" MirrorDB entries (old = if not downloaded after ttl_unaccessed_in_days days, it gets deleted from the table) and dead packages. A package is assumed dead if:
    - after ttl_unaccessed_in_days days it had been updated, it hasn't been actively requested by a client
    - after ttl_unupdated_in_days days in which it had never been requested nor updated. Right now the default value is quite high, set at 300 days (after not updating a package for 300 days with no client asking for it, I consider it dead, so I won't prefetch it anymore)
- Then it starts to download the db files stored in the MirrorDB table into "tmp-db" folder in the pacoloco cache directory, it extract them, parse them and once parsed, stored their relevant entries (package name, arch and version) in RepoPackage table. Then I delete all the "tmp-db" files.
- Then I do an inner join between Package table and RepoPackage table to find only the installed packages which have a different version upstream (in the join condition, i check package name and arch to be equal, and version to be different). Once those have been found, the old packages files (with their signature files) gets deleted and the new version gets prefetched
- The prefetching function is the same handleRequest function adapted to use a single writer (the file one), and it calls a different function to store package data in the db (cause it would have to store different data)
- It supports all the supported packages extensions so far

As future/possible improvements:

- signature checking with prefetched files could be useful 
- Add other heuristics to decide whether discard or prefetch a package

Finally, really thanks for your work so far.

### CODE REVIEW

File: pacoloco.tmpfiles.d [hunk 1]
```diff
@@ -1 +1 @@
-d /var/cache/pacoloco - pacoloco pacoloco
+d /var/cache/pacoloco - pacoloco pacoloco
```
Comment (anatol) [inline pacoloco.tmpfiles.d:1]: it looks like the newline is removed. What is the reason of doing it?

File: README.md [hunk 2]
```diff
@@ -45,19 +55,31 @@ repos:
     url: http://pkgbuild.com/~anatolik/quarry/x86_64
   sublime:
     url: https://download.sublimetext.com/arch/stable/x86_64
+download_prefetch: # optional section, add it if you want to enable prefetching
```
Comment (anatol) [inline README.md:58]: It looks like you want to use crontab-like functionality for the prefetch. Do we really need it? Would a simple interval (e.g. "run prefetch every hour") work here? If not and you really want to use something more powerful then I suggest to look at cron expression parsers like this one https://github.com/gorhill/cronexpr

File: integration_test.go [hunk 3]
```diff
@@ -101,10 +151,12 @@ func testRequestExistingRepo(t *testing.T) {
 	if !os.IsNotExist(err) {
 		t.Error("repo1/test.db should be cached")
 	}
+
 }
 
 func testRequestExistingRepoWithDb(t *testing.T) {
 	// Requesting existing repo
+
```
Comment (anatol) [inline integration_test.go:159]: I see you added a number of newlines. Is it really needed for this pull request?

File: helpers.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+import (
+	"fmt"
+	"log"
+	"os"
+)
+
+func panicIfError(e error) {
+	if e != nil {
+		fmt.Printf("panic:\n%v\n", e)
```
Comment (anatol) [inline helpers.go:11]: I am pretty sure panic will print the error message. No need to print it before panic itself.

File: helpers.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+		panic(e)
+	}
+}
+
+func fatalIfError(err error) {
+	if err != nil {
+		fmt.Printf("fatal error:\n%v\n", err)
```
Comment (anatol) [inline helpers.go:18]: ditto wrt the message

File: helpers.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+)
+
+func panicIfError(e error) {
```
Comment (anatol) [inline helpers.go:9]:
```
if err != nil {
  panic(err)
}
```

Is a standard golang idiom that is a standard way to check for error. I've never seen at the standard library that it moved to a separate function. It does not make much sense to do it in my opinion. So it is just better to inline the function.

File: pacoloco.go [hunk 7]
```diff
@@ -14,17 +14,42 @@ import (
...
 	var err error
 	pathRegex, err = regexp.Compile("^/repo/([^/]*)(/.*)?/([^/]*)$")
-	if err != nil {
-		panic(err)
-	}
+	//source: https://archlinux.org/pacman/makepkg.conf.5.html PKGEXT
+	allowedPackagesExtensions = []string{".pkg.tar.gz", ".pkg.tar.bz2", ".pkg.tar.xz", ".pkg.tar.zst", ".pkg.tar.lzo", ".pkg.tar.lrz", ".pkg.tar.lz4", ".pkg.tar.lz", ".pkg.tar.Z", ".pkg.tar"}
+	panicIfError(err)
+	//This is ugly, explanation here https://regex101.com/r/qB0fQ7/32
+	filenameRegex, err = regexp.Compile("^([a-z0-9._+-]+)-([^-]+-[0-9]+)-([^-]+)(([.]pkg[.]tar(([.]gz)|([.]bz2)|([.]xz)|([.]zst)|([.]lzo)|([.]lrz)|([.]lz4)|([.]lz)|([.]Z))?)([.]sig)?)$")
```
Comment (anatol) [inline pacoloco.go:40]: yeah, these regex are a bit hard to follow. In golang it is sometimes more preferable to do a straight-forward (though sometimes a bit more verbose) solution.

I wonder if it is better just use `strings.HasSuffix()` method to match the extension. Once it is matched - remove it, and split the rest using a regexp or just `strings.Split()`.

Also does this code handle package epochs?

File: pacoloco.go [hunk 8]
```diff
@@ -78,10 +106,79 @@ var (
 	downloadingFilesMutex sync.Mutex
 )
 
+// force resources prefetching
+func prefetchRequest(url string) (err error) {
+
```
Comment (anatol) [inline pacoloco.go:111]: the existing code does not add an empty line after the function name. Let's be consistent with the rest of the code.

**Review by anatol — COMMENTED** [review]
Fantastic Pull Request. Thanks a lot for your work! I really like the amount of tests you added for this code.

I did a first pass of this code review.

A few more notes:

- Please squash all your commits into more coupled commits. The general recommendation is "one feature - one commit". So let's just squash your PR into 1 commit and then you would `--amend` your changes on top of it.
- Run `go fmt` `go vet` over your PR. `revive` tool is another great tool that helps to discover code style issues.

- 

-

File: pacoloco.tmpfiles.d (hunk 1 above)
Comment (Focshole) [reply pacoloco.tmpfiles.d:1]: It's a mistake, nice catch!

File: helpers.go (hunk 4 above)
Comment (Focshole) [reply helpers.go:11]: When running tests I had no error messages or anything readable whatsoever, neither with verbose.
I can remove it if you want, it is not that relevant now

File: helpers.go (hunk 6 above)
Comment (Focshole) [reply helpers.go:9]: Came from other languages, always found it quite verbose and annoying, replace them back would be easy, I'll do it :)

File: README.md (hunk 2 above)
Comment (Focshole) [reply README.md:58]: Imho it is relevant cause we may want prefetching at a specific time cause network quality is also better (e.g. at night the network might be less congestionated). Didn't knew about cronexpr, I'll use it

File: integration_test.go (hunk 3 above)
Comment (Focshole) [reply integration_test.go:159]: Nope, don't remember if it was due to the vscode linter or if i had added them myself

File: pacoloco.go (hunk 7 above)
Comment (Focshole) [reply pacoloco.go:70]: I had thought about being quite strict on what to accept as a string to prefetch, to avoid things getting out of hand (e.g. path traversal/ weird stuff to access files outside of their bounds).
Now I have added a more complete explanation about how do they are matched.
HasSuffix is being used when I check if a file is a signature or a db one. If it is not one of those, I'll check it against the regex

Comment (Focshole) [reply pacoloco.go:70]: Right now I do not know which are the full list of the allowed characters for the package name (this one should be complete, tested against all the dbs in 3 repos, but if I do not find a documentation about it, it won't be bulletproof)
and especially the epoch/version parameter.
The epoch and versions does not get parsed, they are simply treated as strings cause so far I haven't seen a standard formatting. 
I assume the version in the upstream mirror is the most recent one (also cause the prefetching mechanism should mirror what is being kept upstream, otherwise it would be in an incoherent/detached state wrt the db files served by upstream mirrors)

File: pacoloco.go [hunk 9]
```diff
@@ -78,10 +144,78 @@ var (
...
+	} else {
+		return nil
+	}
+}
+
 func handleRequest(w http.ResponseWriter, req *http.Request) error {
 	urlPath := req.URL.Path
 	matches := pathRegex.FindStringSubmatch(urlPath)
-	if matches == nil {
+	if len(matches) == 0 {
```
Comment (Focshole) [inline pacoloco.go:218]: len(matches) should have been used cause if matches is nil, returns 0, but if matches exists and matches nothing, it won't return nil

File: pacoloco.go [hunk 10]
```diff
@@ -246,7 +466,7 @@ func sendCachedFile(w http.ResponseWriter, req *http.Request, fileName string, f
 	if err != nil {
 		return err
 	}
-
+	log.Printf("serving cached file %v", filePath)
```
Comment (Focshole) [inline pacoloco.go:469]: Adding a message when serving a cached file, it was missing (maybe it was intentional?)

File: repo_db_mirror_test.go [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+type testTarDB struct {
+	PkgName string
+	Content string
+}
+
+//testing content for a db
+func getTestTarDB() []testTarDB {
+	return []testTarDB{
+		testTarDB{PkgName: "acl-2.3.1-1",
+			Content: `%FILENAME%
```
Comment (Focshole) [inline repo_db_mirror_test.go:30]: I don't know if there is a better way to do this, e.g. by adding a file which can be easily accessible from tests

File: pacoloco.yaml.sample [hunk 12]
```diff
@@ -1,6 +1,6 @@
 # cache_dir: /var/cache/pacoloco
 # port: 9129
-# download_timeout: 200 # stuck downloads will timeout after 200 sec
+download_timeout: 5 # stuck downloads will timeout after 5 sec
```
Comment (Focshole) [inline pacoloco.yaml.sample:3]: to avoid memory going up too much when multiple large downloads fail

File: README.md (hunk 2 above)
Comment (Focshole) [reply README.md]: I have added it and tested, should be fine now.

**Comment by Focshole** [comment]
I have used `revive` and renamed variables/fixed indentation accordingly, so if you see some variable renaming somewhere outside the prefetching's scope else, it is due to it

**Review by Focshole — COMMENTED** [review]
Now everything should be fine, I have finished doing edits.

**Comment by anatol** [comment]
I'll be glad to add 'Credits' sections to the project. There are many features that were added by community and it is fair to mention it. I am thinking about inserting a widget that shows the data from github 'Contributors' page.

BTW it looks like the email you used for your contribution focshole@notmyemail.com is not associated with your github account. Please check it so github would recognize your accout correctly.

**Comment by anatol** [comment]
https://github.com/anatol/pacoloco/commit/bd7b588ecc122a1ef9d03bd21d134be7ff944fc3 adds 'contributor' widget. Once you register the email at your github account we'll get your name at the widget as well.

**Comment by Focshole** [comment]
Well, this was my fault and I do not think there is a way fix it. I committed with the wrong email so that's it.
I tried with what is being suggested [here](https://stackoverflow.com/a/61368365/4746441) but of course it didn't work cause it is a different repo.
